### PR TITLE
Implement "return to cart" and "place order" buttons on checkout

### LINKED ISSUE #1303: Checkout Block: Form Section - Place Order and Return to Cart actions
As per the exploration done by @Aljullu in #1208 this issue is concerned with the Place Order and Return to cart actions in the Checkout Block Form

In context of entire view:

checkout

* BackLink (defined in #1207)
image
* PlaceOrderButton (`/base/components/checkout/place-order-button`)
image

### PR DESCRIPTION
**Pull request by mikejolley**
Adds a return to cart link, and the place order button.

**Return to cart**
- Can be disabled in block options
- URL can be controlled in block options
- URL defaults to cart page set in WC

**Place order**
- Standard button with styling and callback
- Gets the callback action and text from checkout context
- I have added a TODO for the actual action. The button is non-functional in this PR

**Misc**
Whilst styling these new buttons, I also styled some of the margins on the rest of the form to match the mockup. This was mostly needing to align things vertically.

Fixes #1303

### Screenshots

Screenshot 2020-03-10 at 14 19 34
Screenshot 2020-03-10 at 14 19 17

### How to test the changes in this Pull Request:

1. Go to the checkout
2. Check appearance of link and button
3. Edit block (might be invalidated)
4. Confirm you can toggle the cart link on/off and change the URL.

### CODE REVIEW

**Comment by nerrad** [comment]
> URL defaults to cart page set in WC

I don't think we should address in this pull, but I wonder if we can do some detection of pages that have either the cart block or the default cart page and add that as a dropdown list here for selecting from?

That way we can set a default but the user has a whitelisted set of urls to select from. That will help prevent merchants from shooting themselves in the foot (implementing could be something done in next iteration).

**Comment by nerrad** [comment]
cc @garymurray it'd be good to get your eyes on this too as it's not something in the current designs. (Thanks Rua for the reminder).

**Comment by haszari** [comment]
> URL defaults to cart page set in WC

I'm thinking we use the woo cart page url, and don't give users the option to change it in block settings. This feels like the simplest "happy path" - allowing changing these urls in individual instances of the blocks could easily get confusing, and long term doesn't seem like a useful option.

**Comment by mikejolley** [comment]
> long term doesn't seem like a useful option

Correct. We need a short term solution to keep the cart and checkout flow intact.

**Comment by nerrad** [comment]
> I'm thinking we use the woo cart page url, and don't give users the option to change it in block settings. This feels like the simplest "happy path" - allowing changing these urls in individual instances of the blocks could easily get confusing, and long term doesn't seem like a useful option.

Can you expand on the implications of making this the default Rua? Please include:

- Will merchants have to do any additional configuration elsewhere in the site to get cart and checkout block flow working as expected? If so, what?
- Will merchants be able to keep existing cart and checkout flows intact while they test the cart and checkout blocks?

The solution Mike is proposing here allows for merchants to keep existing cart and checkout flow intact, but experiment with the new block flow on different pages. Keep in mind existing cart and checkout flow is _not_ required by the blocks, they _can_ operate independently of the existing flow, we just need to ensure that the flow between them can be adjusted. Mike's pull allows for that (as a short-term accommodation for initial release). I think this can make introducing the new blocks a little bit less unnerving for folks and _might_ get us more *initial* testing because it's not an _either/or_ approach. 

Once we have some actionable feedback after the blocks are released _then_ we can start looking at implementing tighter flows/automatic setup for wider adoption (and better out of the box experience) of the new blocks (eventually _replacing_ the shortcodes).

**Comment by haszari** [comment]
> Can you expand on the implications of making this the default Rua?

I haven't explored this in full - would need to spend some time working it through to understand all the potential implications. I figure it's better for a designer to do that, but I could explore more if that's helpful.

**Comment by haszari** [comment]
There's one issue that I hit yesterday that made me think there might be benefits to working within the current cart & checkout pages - logged here: https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/1930

File: assets/js/base/context/checkout-context.js [hunk 1]
```diff
@@ -55,7 +58,7 @@ const CheckoutProvider = ( {
 		checkoutHasError,
 		activePaymentMethod,
 		placeOrderLabel,
-		notices,
+		placeOrderAction.notices,
```
Comment (haszari) [inline assets/js/base/context/checkout-context.js]: Will this be undefined? (Since `placeOrderAction` is an empty placeholder function above)

I didn't see any JS errors, so not sure if this code is firing.

**Comment by garymurray** [comment]
@mikejolley @nerrad @haszari thinking through a few scenarios here (there might be more I miss):

**1. Merchants keep the current shortcode based cart and checkout published and adds new pages with the block-based cart and checkout to test/compare**

In this situation - which checkout will the store default too? 

If the merchant wanted to actually test the new checkout, would they not have to set the add-to-cart buttons to take customers to the new checkout flow? 

Should we rather have a setting (or piggyback on the existing page template setting in WC Core) that allows merchants to chose which checkout they would like to use for checkout - the old or new - and then the urls update accordingly? This seems to be getting complex though.

**2. Merchant sets the current shortcode based cart and checkout pages to draft and publishes new pages with the block-based cart and checkout**

Can the urls used for the add-to-cart, return to cart etc automatically update to the new urls?

Are there scenarios I am missing?

**Comment by mikejolley** [comment]
@garymurray When you say "default to" I assume you mean where things in themes link to? In which case, any existing cart/checkout links go to the assigned page in WC > Settings. In this scenario I assume it would be the shortcode versions unless the merchant specifically wants to switch to new.

2 doesn't matter. It would be the same for the merchant; go to WC > Settings and change the cart/checkout page to the "new" pages.

Screenshot 2020-03-11 at 14 37 38

To clarify, the issue we have here is that unless you set those ^ to the new block based cart/checkout, the block based cart/checkout have no knowledge of what URL their sibling lives. Each page is independent.

File: assets/js/base/context/checkout-context.js (hunk 1 above)
Comment (mikejolley) [reply assets/js/base/context/checkout-context.js]: Will be fixed after master rebase.

File: assets/js/base/context/checkout-context.js [hunk 2]
```diff
@@ -19,6 +19,8 @@ const CheckoutProvider = ( {
 	children,
 	initialActivePaymentMethod,
 	placeOrderLabel = __( 'Place Order', 'woo-gutenberg-product-block' ),
+	// @todo Implement usable action for place order.
```
[the remark this answers is not in the document]
Comment (mikejolley) [reply assets/js/base/context/checkout-context.js]: This is already being worked on by @nerrad and is part of the new context for checkout.

**Comment by mikejolley** [comment]
As discussed, this now includes a select box for the page. cc @nerrad 

Screenshot 2020-03-11 at 15 40 27

If approved we can roll out the same option in the cart block.

**Comment by garymurray** [comment]
After chatting with @mikejolley this morning p1584002713264800-slack-woo-merchandising I am happy for us to go ahead with this option to allow merchants to link up the blocks to test them while their site continues to use the current shortcode/default cart and checkout pages.

As Mike also explained - we can hide this option if these new pages are already setup as the new default cart and checkout pages in the WC Settings.

I do think we should include an option though which prompts our merchants to go change the default cart and checkout pages if they insert these blocks on new pages - so - my question would be - if someone adds the cart and checkout blocks to new pages, can we conditional show a notice like this:

Screenshot 2020-03-12 at 11 34 29

If the merchant is inserting these blocks into the existing pages as set in the WC settings - then that message should not show.  Is this possible?

**Comment by LevinMedia** [comment]
@garymurray I like the reminder notice, but I don't think we have to make it dismissible. I think it's a good reminder, regardless, as long as we check that the page you're editing isn't already set as the cart or checkout page.

**Comment by mikejolley** [comment]
This is ready for final review.

Added the notice suggested above if the page is not assigned as the woo checkout page:
Screenshot 2020-03-13 at 15 39 30

Updated the select control to include a default (Woo cart page), and this is hidden if left to default and the checkout page is assigned as the real woo checkout page in settings:
Screenshot 2020-03-13 at 15 39 19

I had to extract the feedback component to move it so we didn't have lots of notices next to each other.

**Comment by haszari** [comment]
@mikejolley I'm holding off re-testing this until it's rebased. I had a quick look and the page picker didn't seem to render for me:

Screen Shot 2020-03-16 at 10 11 31 AM

Regarding `ALL_PAGES` - do we need to enqueue this data? I'm wondering if we can pull the page list from a core API/store, e.g. something like `useSelect( 'core/pages' )`.

File: assets/js/blocks/cart-checkout/checkout/edit.js [hunk 3]
```diff
@@ -200,39 +239,93 @@ const CheckoutEditor = ( { attributes, setAttributes } ) => {
...
+						checked={ showReturnToCart }
+						onChange={ () =>
+							setAttributes( {
+								showReturnToCart: ! showReturnToCart,
+							} )
+						}
+					/>
+					{ showReturnToCart &&
+						( currentPostId !== CHECKOUT_PAGE_ID ||
+							cartPageId ) && (
```
Comment (haszari) [inline assets/js/blocks/cart-checkout/checkout/edit.js]: Is it intended that the page picker is displayed only if:

- The checkout block is on the Woo `checkout` page (at the time of editing).
- OR
- The user has previously set a cart page id.

In the case that the merchant has set up both woo `cart` & `checkout` pages with blocks, but previously was using shortcodes, and had set a `cartPageId` (to test blocks) .. I think the page picker would display? Probably an edge case, I'm thinking maybe we can don't need the `|| cartPageId`.

Comment (mikejolley) [reply assets/js/blocks/cart-checkout/checkout/edit.js]: > The checkout block is on the Woo checkout page (at the time of editing)

Backwards. Picker if the page is NOT the woo checkout page at time of editing. 

The picker displays because they have made a selection, so we continue to respect their choice. It avoids extra logic when reading that value during block rendering too - otherwise it would need to decide whether or not to use the option which is too much effort imo

**Comment by mikejolley** [comment]
Endless rebasing yay

File: assets/js/blocks/cart-checkout/checkout/block.js [hunk 4]
```diff
@@ -324,6 +329,18 @@ const Block = ( {
 								}
 							/>
 						</FormStep>
+						<div className="wc-block-checkout__actions">
+							{ attributes.showReturnToCart && (
+								<ReturnToCartButton
+									link={
+										getSetting(
+											'page-' + attributes.cartPageId
+										) || CART_URL
```
Comment (nerrad) [inline assets/js/blocks/cart-checkout/checkout/block.js]: `getSetting` can have a default, maybe move `CART_URL` in as the default here? Alternatively, the `ReturnToCartButton` _already_ uses `CART_URL` if the link prop is undefined or an empty string... so maybe just pass in the result of the `getSetting` call?

File: assets/js/blocks/cart-checkout/checkout/block.js [hunk 5]
```diff
@@ -333,9 +333,12 @@ const Block = ( {
...
 									link={
-										getSetting(
-											'page-' + attributes.cartPageId
-										) || CART_URL
+										attributes.cartPageId
+											? getSetting(
+													'page-' +
+														attributes.cartPageId
+											  )
+											: false
```
Comment (nerrad) [inline assets/js/blocks/cart-checkout/checkout/block.js]: This could be simplified to:

```suggestion
										 getSetting(
													'page-' +
														attributes?.cartPageId,
													false
											  )
```

Comment (mikejolley) [reply assets/js/blocks/cart-checkout/checkout/block.js]: Isn't this an unnecessary lookup into the settings array though?

Comment (nerrad) [reply assets/js/blocks/cart-checkout/checkout/block.js]: Not really expensive imo. The array is not large.
